src/libs/history.c

```diff
@@ -184,6 +250,24 @@ static void _lib_history_change_callback(gpointer instance, gpointer user_data)
                      TRUE, TRUE, 0);
   num++;
 
+  if (d->record_undo == TRUE)
+  {
+    /* record undo/redo history snapshot */
+    if (d->prev.snapshot != NULL)
+    {
+      dt_undo_history_t *hist = malloc(sizeof(dt_undo_history_t));
+      hist->snapshot=d->prev.snapshot;
+      hist->end=d->prev.end;
+      dt_undo_record(darktable.undo, self, DT_UNDO_HISTORY, (dt_undo_data_t *)hist, &pop_undo, _history_undo_data_free);
+    }
+
+    // record current history for next iteration, we duplicate the whole history
+    d->prev.snapshot = _duplicate_history(darktable.develop->history);
```

Comment:
And what if this function (`_lib_history_change_callback()`) is called only once, so `d->prev.snapshot` will not get stored in `dt_undo_record()`?
It will leak.

So one of the solutions is to add this into `gui_cleanup()`:
```
dt_lib_history_t *d = (dt_lib_history_t *)self->data;
g_list_free_full(d->prev.snapshot, dt_dev_free_history_item);
```
(before `g_free(self->data);`)